x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java

```diff
@@ -6,42 +6,59 @@
 
 package org.elasticsearch.xpack.runtimefields.mapper;
 
-import org.apache.lucene.document.FieldType;
 import org.apache.lucene.util.SetOnce;
 import org.elasticsearch.common.xcontent.XContentBuilder;
-import org.elasticsearch.common.xcontent.support.XContentMapValues;
 import org.elasticsearch.index.mapper.FieldMapper;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.Mapper;
 import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.ParametrizedFieldMapper;
 import org.elasticsearch.index.mapper.ParseContext;
+import org.elasticsearch.index.mapper.TypeParsers;
+import org.elasticsearch.painless.PainlessScriptEngine;
 import org.elasticsearch.script.Script;
 import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.ScriptType;
 import org.elasticsearch.xpack.runtimefields.StringScriptFieldScript;
 
 import java.io.IOException;
-import java.util.Iterator;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-public final class ScriptFieldMapper extends FieldMapper {
+public final class ScriptFieldMapper extends ParametrizedFieldMapper {
 
     public static final String CONTENT_TYPE = "script";
 
-    private static final FieldType FIELD_TYPE = new FieldType();
+    private final String runtimeType;
+    private final Script script;
 
-    ScriptFieldMapper(String simpleName, MappedFieldType mappedFieldType, MultiFields multiFields, CopyTo copyTo) {
-        super(simpleName, FIELD_TYPE, mappedFieldType, multiFields, copyTo);
+    private static ScriptFieldMapper toType(FieldMapper in) {
+        return (ScriptFieldMapper) in;
+    }
+
+    protected ScriptFieldMapper(
+        String simpleName,
+        MappedFieldType mappedFieldType,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        Builder builder
+    ) {
+        super(simpleName, mappedFieldType, multiFields, copyTo);
+        // TODO is it ok that the object being built needs to read from the object that is building it? Shouldn't Builder#build return
+        // a complete object? Maybe all the parameters need to be passed through instead of the whole builder?
```

Comment:
@romseygeek this is for you, does it make sense?

Reply:
Given that the mapper constructor is protected here, I think it's OK?  You can pull out individual parameters if you think it's clearer, I used this pattern on other mappers simply because otherwise you end up with 14 constructor parameters most of which are just boilerplate.